File: Chess/src/main/java/game/pieces/King.java
package game.pieces;


import game.boards.Board;
import game.boards.Tile;
import game.move.Move;
import game.resources.PIECE_Configurations;
import game.util.BoardUtilities;

import java.util.ArrayList;
import java.util.List;


public class King extends Piece {

    private boolean castlingDone = false;

    public King(Team team) {
        super(team, PieceTypes.KING);
    }

    public boolean isCastlingDone() {
        return castlingDone;
    }

    public void setCastlingDone(boolean castlingDone) {
        this.castlingDone = castlingDone;
    }

    @Override
    public List<Move> availableMoves(Board board, Coordinate currentCoord) {
        List<Move> possibleMoves = new ArrayList<Move>();
        Tile currentTile = board.getTile(currentCoord);
        Tile destinationTile;
        Coordinate destinationCoordinate;
        for (Coordinate coord : PIECE_Configurations.QUUEN_MOVES) {
            destinationCoordinate = currentCoord.plus(coord);
            if (!BoardUtilities.isValidCoordinate(destinationCoordinate)) {
                continue;
            }
            destinationTile = board.getTile(destinationCoordinate);
            if (!destinationTile.hasPiece()) {
                possibleMoves.add(new Move(board, currentTile, destinationTile));
            } else {
                if (destinationTile.getPiece().getTeam() != this.getTeam()) {
                    possibleMoves.add(new Move(board, currentTile, destinationTile));
                }
            }
        }

        return possibleMoves;

    }

}

File: Chess/src/main/kotlin/hummel/game/piece/King.kt
package main.kotlin.hummel.game.piece

import main.kotlin.hummel.game.Move
import main.kotlin.hummel.game.Pieces
import main.kotlin.hummel.game.board.Board
import main.kotlin.hummel.game.board.Tile
import main.kotlin.hummel.game.util.BoardUtilities

class King(team: Team) : Piece(team, PieceTypes.KING) {
	override fun availableMoves(board: Board, currentCoord: Coordinate): List<Move> {
		val possibleMoves: MutableList<Move> = ArrayList()
		val currentTile = board.getTile(currentCoord)
		var destinationTile: Tile
		var destinationCoordinate: Coordinate?
		for (coord in Pieces.QUEEN_MOVES) {
			destinationCoordinate = currentCoord.plus(coord)
			if (!BoardUtilities.isValidCoordinate(destinationCoordinate)) {
				continue
			}
			destinationTile = board.getTile(destinationCoordinate)
			if (!destinationTile.hasPiece()) {
				Move(board, currentTile, destinationTile).let { possibleMoves.add(it) }
			} else {
				if (destinationTile.piece.team !== team) {
					Move(board, currentTile, destinationTile).let { possibleMoves.add(it) }
				}
			}
		}
		return possibleMoves
	}
}